### PR TITLE
Fall back to using label for preferred devices

### PR DESCRIPTION
**Pull request by damencho**
This brings back the changes from https://github.com/jitsi/jitsi-meet/pull/3219. Some of the changes were introduced earlier and this adopts the PR using the recently introduced new storage values for user selected device id and adds a label to them.

### CODE REVIEW

File: react/features/base/devices/functions.js [hunk 1]
```diff
@@ -67,6 +67,34 @@ export function getDeviceIdByLabel(state: Object, label: string, kind: string) {
     }
 }
 
+/**
+ * Finds a device with a label that matches the passed id and returns its label.
```
Comment (virtuacoplenny) [inline react/features/base/devices/functions.js]: Side musing: this makes me wonder if we even need to store deviceId at all and instead can rely solely on label. Based on what you've seen would it be better to move to label over device id? There would likely need to be a translation layer, maybe in lib-jitsi-meet to convert label to id for the sake of calling gUM with an exact device id.

File: react/features/base/devices/functions.js [hunk 2]
```diff
@@ -118,24 +146,29 @@ export function groupDevicesByKind(devices: Object[]): Object {
...
-        userSelection: boolean = false): Promise<*> {
+        userSelection: boolean = false,
+        newLabel: ?string): Promise<*> {
     return JitsiMeetJS.mediaDevices.setAudioOutputDevice(newId)
         .then(() => {
             const newSettings = {
                 audioOutputDeviceId: newId,
-                userSelectedAudioOutputDeviceId: undefined
+                userSelectedAudioOutputDeviceId: undefined,
+                userSelectedAudioOutputDeviceLabel: undefined
```
Comment (virtuacoplenny) [inline react/features/base/devices/functions.js]: Why does this need to be explicitly set to undefined?

File: react/features/base/settings/functions.js [hunk 3]
```diff
@@ -97,3 +97,152 @@ export function getServerURL(stateful: Object | Function) {
...
+    // If there is no label at all, there is no need to fall back to checking
+    // the label for a fuzzy match.
+    if (!userSelectedDeviceLabel || !userSelectedDeviceId) {
+        return userSelectedDeviceId;
+    }
+
+    const foundMatchingBasedonDeviceId = availableDevices.find(
+        candidate => candidate.deviceId === userSelectedDeviceId);
+
+    // Prioritize matching the deviceId of the microphone.
```
Comment (virtuacoplenny) [inline react/features/base/settings/functions.js]: Specific mention of microphone not needed.

File: react/features/device-selection/components/AudioOutputPreview.js [hunk 4]
```diff
@@ -113,6 +113,7 @@ class AudioOutputPreview extends Component<Props> {
      */
     _setAudioSink() {
         this._audioElement
+            && this.props.deviceId
```
Comment (virtuacoplenny) [inline react/features/device-selection/components/AudioOutputPreview.js]: Should we disable play if there is no device id? I'm not sure what the behavior of <Audio /> is if we don't set a sink id.

File: conference.js [hunk 5]
```diff
@@ -2277,20 +2288,23 @@ export default {
...
 
     /**
-     * Inits list of current devices and event listener for device change.
+     * Updates the list of current devices.
+     * @param {boolean} setDeviceListChangeHandler - Whether to add the deviceList change handlers.
      * @private
      * @returns {Promise}
      */
-    _initDeviceList() {
+    _initDeviceList(setDeviceListChangeHandler = false) {
```
[the remark this answers is not in the document]
Comment (damencho) [reply conference.js]: Doesn't matter this code should, I made it like this minimizing changes. And anyway we need to refactor this code later moving everything to react ...

File: conference.js [hunk 6]
```diff
@@ -718,13 +718,21 @@ export default {
         this.roomName = options.roomName;
 
         return (
-            this.createInitialLocalTracksAndConnect(
+
+            // Initialize the device list first. This way, when creating tracks
+            // based on preferred devices, loose label matching can be done in
+            // cases where the exact ID match is no longer available, such as
+            // when the camera device has switched USB ports.
+            this._initDeviceList()
```
[the remark this answers is not in the document]
Comment (damencho) [reply conference.js]: I will test those cases on Tuesday.

File: react/features/base/devices/functions.js (hunk 1 above)
Comment (damencho) [reply react/features/base/devices/functions.js]: Not a change for now, if we decide to do it ...

File: react/features/base/devices/functions.js (hunk 2 above)
Comment (damencho) [reply react/features/base/devices/functions.js]: It is mentioned why a few lines below. This is flow workaround with the delete in the else below. Couldn't fix the flow error another way.
image

File: react/features/device-selection/components/AudioOutputPreview.js (hunk 4 above)
Comment (damencho) [reply react/features/device-selection/components/AudioOutputPreview.js]: It fallbacks to default device I think, this is what I see when playing after not setting it. There is no error or anything.

File: react/features/base/devices/functions.js (hunk 1 above)
Comment (virtuacoplenny) [reply react/features/base/devices/functions.js]: That wasn't the question though.

Comment (damencho) [reply react/features/base/devices/functions.js]: ?

**Pull request merged**